File: clients/go/signature.go
// Verifying that a webhook came from Hook0, and that nothing in it changed on the way.
//
// A signature header names the moment it was signed and one or two message authentication codes
// over the body. The `v1` scheme also covers a list of request headers, so a receiver can tell apart
// two deliveries that carry the same body but not the same context; `v0` covers the body alone and
// is what an older sender still produces. When both are offered, `v1` is the one verified: accepting
// the weaker of two schemes on the strength of the sender offering it is how a downgrade works.
//
// Two things are refused before any code is computed. A header the signature says it covers but the
// request did not carry is refused outright, because signing over an absent value would let a sender
// drop a header and keep the signature valid. And a signature whose codes are not whole hexadecimal
// is refused rather than decoded as far as it goes: a decoder that stops at the first bad character
// compares a prefix, and a prefix of the right code is not the right code.
//
// The clock window is bilateral. A moment too far in the future is refused exactly like one too far
// in the past, so the window a given delivery is accepted in stays the width the caller asked for,
// whichever way a clock drifted.

package hook0

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"
)

const (
	// maxSignatureBytes is the longest signature header read. The header is written by whoever
	// reached the endpoint, so its size is bounded before any of it is split, decoded or compared.
	maxSignatureBytes = 8 * 1024

	// maxSignatureParts is the most `key=value` parts one signature header is split into.
	maxSignatureParts = 32

	// maxCoveredHeaders is the most header names one signature covers.
	maxCoveredHeaders = 64

	// maxTimestampSeconds is the furthest from the epoch, in either direction, a signature's moment
	// may sit. Holding a moment thousands of years out against the current time is arithmetic no
	// caller expects to be doing.
	maxTimestampSeconds = 1_000_000_000_000

	// partSeparator is what separates one part of the signature header from the next.
	partSeparator = ","

	// partAssignator is what separates the name of a part from its value. Only the first one
	// counts: a value may hold further ones, and splitting on all of them would silently drop
	// everything past the second.
	partAssignator = "="

	// headerNameSeparator is what separates two header names inside the `h` part, and what they are
	// joined back with.
	headerNameSeparator = " "

	// messageSeparator is what separates the pieces of the message a code is computed over.
	messageSeparator = "."

	// timestampPart names the moment the delivery was signed, in whole seconds since the epoch.
	timestampPart = "t"

	// bodySchemePart carries the code covering the body alone.
	bodySchemePart = "v0"

	// headersSchemePart carries the code covering the covered headers and the body.
	headersSchemePart = "v1"

	// coveredHeadersPart lists the headers the `v1` code covers, in the order it covers them.
	coveredHeadersPart = "h"

	// headerNameCharacters is what a header name is written with, as RFC 9110 spells a token.
	headerNameCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!#$%&'*+-.^_`|~"
)

// Signature is a signature header, read into the pieces a verification needs.
type Signature struct {
	// Timestamp is the moment the delivery was signed, in whole seconds since the epoch.
	Timestamp int64
	// CoveredHeaders names the headers the stronger scheme covers, in the order it covers them and
	// lowercased.
	CoveredHeaders []string
	// BodyCode is the `v0` code, nil when the signature offers none.
	BodyCode []byte
	// HeadersCode is the `v1` code, nil when the signature offers none.
	HeadersCode []byte
}

// ParseSignature reads a signature header, refusing anything it cannot read whole.
func ParseSignature(signature string) (*Signature, error) {
	if len(signature) > maxSignatureBytes {
		return nil, fmt.Errorf(
			"%w: it is %d characters long, above the %d accepted",
			ErrSignatureUnreadable, len(signature), maxSignatureBytes,
		)
	}

	parts := strings.Split(signature, partSeparator)
	if len(parts) > maxSignatureParts {
		return nil, fmt.Errorf(
			"%w: it carries more than the %d parts accepted",
			ErrSignatureUnreadable, maxSignatureParts,
		)
	}

	read := make(map[string]string, len(parts))
	for _, part := range parts {
		name, value, found := strings.Cut(part, partAssignator)
		if !found {
			continue
		}
		read[strings.TrimSpace(name)] = strings.TrimSpace(value)
	}
	if len(read) < 2 {
		return nil, fmt.Errorf("%w: it carries neither a moment nor a code", ErrSignatureUnreadable)
	}

	timestamp, err := signatureTimestamp(read)
	if err != nil {
		return nil, err
	}
	covered, err := coveredHeaders(read)
	if err != nil {
		return nil, err
	}
	bodyCode, err := signatureCode(read, bodySchemePart)
	if err != nil {
		return nil, err
	}
	headersCode, err := signatureCode(read, headersSchemePart)
	if err != nil {
		return nil, err
	}
	if bodyCode == nil && headersCode == nil {
		return nil, fmt.Errorf(
			"%w: it carries neither a `%s` nor a `%s` code",
			ErrSignatureUnreadable, bodySchemePart, headersSchemePart,
		)
	}

	return &Signature{
		Timestamp:      timestamp,
		CoveredHeaders: covered,
		BodyCode:       bodyCode,
		HeadersCode:    headersCode,
	}, nil
}

// Verify reports whether the code this signature carries is the one the secret produces.
//
// The stronger scheme wins when both are offered, and the comparison is made in constant time: one
// that gave up at the first differing byte would say, by how long it took, how much of a guess was
// right.
func (s *Signature) Verify(payload []byte, coveredValues []string, subscriptionSecret string) bool {
	code := hmac.New(sha256.New, []byte(subscriptionSecret))
	code.Write([]byte(strconv.FormatInt(s.Timestamp, 10)))
	code.Write([]byte(messageSeparator))

	if s.HeadersCode != nil {
		code.Write([]byte(strings.Join(s.CoveredHeaders, headerNameSeparator)))
		code.Write([]byte(messageSeparator))
		code.Write([]byte(strings.Join(coveredValues, messageSeparator)))
		code.Write([]byte(messageSeparator))
		code.Write(payload)
		return hmac.Equal(code.Sum(nil), s.HeadersCode)
	}

	if s.BodyCode != nil {
		code.Write(payload)
		return hmac.Equal(code.Sum(nil), s.BodyCode)
	}

	// Unreachable: a signature carrying neither code is refused while it is being read.
	return false
}

// signatureTimestamp answers the moment the signature names, which it is not a signature without.
func signatureTimestamp(read map[string]string) (int64, error) {
	written, carried := read[timestampPart]
	if !carried {
		return 0, fmt.Errorf("%w: it carries no `%s` part", ErrSignatureUnreadable, timestampPart)
	}

	seconds, err := strconv.ParseInt(written, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("%w: `%s` is not a number of seconds", ErrSignatureUnreadable, written)
	}
	if seconds > maxTimestampSeconds || seconds < -maxTimestampSeconds {
		return 0, fmt.Errorf(
			"%w: its moment is further than %d seconds from the epoch",
			ErrSignatureUnreadable, maxTimestampSeconds,
		)
	}
	return seconds, nil
}

// signatureCode answers one of the codes a signature offers, decoded whole or not at all.
func signatureCode(read map[string]string, part string) ([]byte, error) {
	written, carried := read[part]
	if !carried {
		return nil, nil
	}

	decoded, err := hex.DecodeString(written)
	if err != nil || len(decoded) == 0 {
		return nil, fmt.Errorf("%w: the `%s` code is not hexadecimal", ErrSignatureUnreadable, part)
	}
	return decoded, nil
}

// coveredHeaders answers the headers the stronger scheme covers, in the order it covers them.
func coveredHeaders(read map[string]string) ([]string, error) {
	written, carried := read[coveredHeadersPart]
	if !carried || written == "" {
		return nil, nil
	}

	names := strings.Split(written, headerNameSeparator)
	if len(names) > maxCoveredHeaders {
		return nil, fmt.Errorf(
			"%w: it covers more than the %d headers accepted",
			ErrSignatureUnreadable, maxCoveredHeaders,
		)
	}

	covered := make([]string, 0, len(names))
	for _, name := range names {
		if name == "" || strings.ContainsFunc(name, isNotHeaderNameCharacter) {
			return nil, fmt.Errorf("%w: `%s` is not a header name", ErrSignatureUnreadable, name)
		}
		covered = append(covered, strings.ToLower(name))
	}
	return covered, nil
}

// isNotHeaderNameCharacter reports whether a character has no place in a header name.
func isNotHeaderNameCharacter(character rune) bool {
	return !strings.ContainsRune(headerNameCharacters, character)
}

// delivered answers the headers of the request under the names a signature refers to them by.
//
// A later value wins over an earlier one under the same name, which is what a map built by the
// caller would have done.
func delivered(headers http.Header) map[string]string {
	carried := make(map[string]string, len(headers))
	for name, values := range headers {
		if len(values) == 0 {
			continue
		}
		carried[strings.ToLower(name)] = values[len(values)-1]
	}
	return carried
}

// VerifyWebhookSignatureAt verifies a webhook against a moment the caller names.
//
//   - signature: the value of the `X-Hook0-Signature` header.
//   - payload: the raw body of the webhook request.
//   - headers: the headers of the webhook request.
//   - subscriptionSecret: the signing secret of the subscription the webhook was delivered for.
//   - tolerance: how far, in either direction, the moment the signature names may sit from
//     currentTime. Five minutes is a reasonable trade-off between tolerating clock drift and
//     bounding how long a captured delivery can be replayed.
//   - currentTime: what to hold the signature's moment against.
//
// Every reason a webhook is refused is one of the sentinels this package declares, so errors.Is
// tells a missing header from a code that does not match from a moment out of the window.
func VerifyWebhookSignatureAt(
	signature string,
	payload []byte,
	headers http.Header,
	subscriptionSecret string,
	tolerance time.Duration,
	currentTime time.Time,
) error {
	parsed, err := ParseSignature(signature)
	if err != nil {
		return err
	}

	carried := delivered(headers)
	coveredValues := make([]string, 0, len(parsed.CoveredHeaders))
	for _, name := range parsed.CoveredHeaders {
		value, found := carried[name]
		if !found {
			return fmt.Errorf("%w: `%s`", ErrHeaderNotDelivered, name)
		}
		coveredValues = append(coveredValues, value)
	}

	if !parsed.Verify(payload, coveredValues, subscriptionSecret) {
		return ErrSignatureMismatch
	}

	drift := currentTime.Sub(time.Unix(parsed.Timestamp, 0))
	if drift < 0 {
		drift = -drift
	}
	if drift > tolerance {
		return fmt.Errorf(
			"%w: it was made %s from now, outside the %s accepted",
			ErrSignatureOutsideTolerance, drift, tolerance,
		)
	}

	return nil
}

// VerifyWebhookSignature verifies a webhook against the current moment.
//
// See VerifyWebhookSignatureAt for what each argument is.
func VerifyWebhookSignature(
	signature string,
	payload []byte,
	headers http.Header,
	subscriptionSecret string,
	tolerance time.Duration,
) error {
	return VerifyWebhookSignatureAt(signature, payload, headers, subscriptionSecret, tolerance, time.Now())
}

File: clients/zig/src/signature.zig
//! Verifying that a webhook came from Hook0, and that nothing in it changed on the way.
//!
//! A signature names the moment it was signed and one or two message authentication codes over the
//! body. The `v1` scheme also covers a list of request headers, so a receiver can tell apart two
//! deliveries that carry the same body but not the same context; `v0` covers the body alone and is
//! what an older sender still produces. When both are offered, `v1` is the one verified: accepting
//! the weaker of two schemes on the strength of the sender offering it is how a downgrade works.
//!
//! Two things are refused before any code is computed. A header the signature says it covers but the
//! request did not carry is refused outright, because signing over an absent value would let a sender
//! drop a header and keep the signature valid. And a signature whose codes are not whole hexadecimal
//! is refused rather than decoded as far as it goes: a decoder that stops at the first bad character
//! compares a prefix, and a prefix of the right code is not the right code.
//!
//! Nothing here allocates. A signature is read into a fixed-size value on the stack, the codes are
//! decoded into fixed arrays, and the message a code is computed over is fed to the keyed hash in
//! pieces rather than joined first — so what a hostile header can cost is settled by this file's own
//! ceilings rather than by an allocator.

const std = @import("std");

const Hmac = std.crypto.auth.hmac.sha2.HmacSha256;

/// Every way a delivery is refused.
///
/// One error per way rather than one error and a string: a caller that has to tell a replayed
/// delivery from a forged one matches on the error, and the four the shared conformance corpus names
/// are the four here.
pub const SignatureError = error{
    /// The header carries a code that is not whole hexadecimal.
    CodeNotHexadecimal,
    /// The signature covers a header the request did not carry.
    HeaderNotDelivered,
    /// The code is not the one the subscription secret produces.
    CodeMismatch,
    /// The moment the signature names sits outside the window the caller allowed.
    OutsideTolerance,
    /// The header is not one this client can read at all.
    Unreadable,
};

/// Longest signature header read. The header is written by whoever reached the endpoint, so its size
/// is bounded before any of it is split, decoded or compared.
pub const max_signature_bytes: usize = 8 * 1024;

/// Most `key=value` parts one signature header is split into.
pub const max_signature_parts: usize = 32;

/// Most header names one signature covers.
pub const max_covered_headers: usize = 64;

/// Furthest from the epoch, in either direction, a signature's moment may sit.
pub const max_timestamp: i64 = 1_000_000_000_000;

/// What separates one part of the signature header from the next.
pub const part_separator: u8 = ',';

/// What separates the name of a part from its value. Only the first one counts: a value may hold
/// further ones, and splitting on all of them would silently drop everything past the second.
pub const part_assignator: u8 = '=';

/// What separates two header names inside the `h` part, and what they are joined back with.
pub const header_name_separator: u8 = ' ';

/// What separates the pieces of the message a code is computed over.
pub const message_separator: u8 = '.';

/// Part naming the moment the delivery was signed, in whole seconds since the Unix epoch.
pub const timestamp_part = "t";

/// Part carrying the code covering the body alone.
pub const body_scheme_part = "v0";

/// Part carrying the code covering the covered headers and the body.
pub const headers_scheme_part = "v1";

/// Part listing the headers the `v1` code covers, in the order it covers them.
pub const covered_headers_part = "h";

/// One header of a delivery, as it arrived.
pub const Header = struct {
    name: []const u8,
    value: []const u8,
};

/// A signature header, read into the pieces a verification needs.
pub const Signature = struct {
    /// The moment the delivery was signed, in whole seconds since the epoch.
    timestamp: i64,
    /// The headers the stronger scheme covers, lowercased and in the order it covers them.
    covered_headers: [max_covered_headers][]const u8 = undefined,
    covered_count: usize = 0,
    /// The `v0` code, decoded, when the header offered one.
    body_code: ?[Hmac.mac_length]u8 = null,
    /// The `v1` code, decoded, when the header offered one.
    headers_code: ?[Hmac.mac_length]u8 = null,

    /// The headers this signature covers, as the slice it filled in.
    pub fn covered(self: *const Signature) []const []const u8 {
        return self.covered_headers[0..self.covered_count];
    }
};

/// The text either side of the first assignator, and nothing when there is not one.
fn partitioned(part: []const u8) ?struct { name: []const u8, value: []const u8 } {
    const at = std.mem.indexOfScalar(u8, part, part_assignator) orelse return null;
    return .{ .name = part[0..at], .value = part[at + 1 ..] };
}

fn trimmed(text: []const u8) []const u8 {
    return std.mem.trim(u8, text, " \t");
}

/// What a header name is written with, as RFC 9110 spells a token.
fn isHeaderName(name: []const u8) bool {
    if (name.len == 0) return false;
    for (name) |character| {
        const usable = switch (character) {
            'A'...'Z', 'a'...'z', '0'...'9' => true,
            '!', '#', '$', '%', '&', '\'', '*', '+', '-', '.', '^', '_', '`', '|', '~' => true,
            else => false,
        };
        if (!usable) return false;
    }
    return true;
}

/// One of the codes a signature offers, decoded whole or not at all.
///
/// A code of the wrong length, an odd number of digits or a character that is not hexadecimal are
/// all one refusal: a decoder that stopped at the first of them would compare a prefix.
fn codeOf(written: []const u8) SignatureError![Hmac.mac_length]u8 {
    if (written.len == 0 or written.len % 2 != 0) return error.CodeNotHexadecimal;
    for (written) |character| {
        if (std.fmt.charToDigit(character, 16) catch null == null) return error.CodeNotHexadecimal;
    }
    if (written.len != Hmac.mac_length * 2) return error.CodeNotHexadecimal;

    var decoded: [Hmac.mac_length]u8 = undefined;
    _ = std.fmt.hexToBytes(&decoded, written) catch return error.CodeNotHexadecimal;
    return decoded;
}

/// The moment the signature names, which it is not a signature without.
fn timestampOf(written: []const u8) SignatureError!i64 {
    if (written.len == 0) return error.Unreadable;

    const digits = if (written[0] == '-') written[1..] else written;
    if (digits.len == 0) return error.Unreadable;
    for (digits) |character| {
        if (character < '0' or character > '9') return error.Unreadable;
    }

    const seconds = std.fmt.parseInt(i64, written, 10) catch return error.Unreadable;
    if (seconds > max_timestamp or seconds < -max_timestamp) return error.Unreadable;
    return seconds;
}

/// Reads a signature header, refusing anything it cannot read whole.
pub fn parse(header: []const u8) SignatureError!Signature {
    if (header.len > max_signature_bytes) return error.Unreadable;

    var read: Signature = .{ .timestamp = 0 };
    var moment: ?[]const u8 = null;
    var covered: ?[]const u8 = null;
    var body: ?[]const u8 = null;
    var headers: ?[]const u8 = null;
    var named: usize = 0;

    var parts = std.mem.splitScalar(u8, header, part_separator);
    var seen: usize = 0;
    while (parts.next()) |part| {
        seen += 1;
        if (seen > max_signature_parts) return error.Unreadable;

        const split = partitioned(part) orelse continue;
        const name = trimmed(split.name);
        const carried = trimmed(split.value);

        // What is counted is the parts a signature names rather than the parts it carries: a header
        // naming one of them twice says one thing, not two, and the second value is the one kept.
        if (std.mem.eql(u8, name, timestamp_part)) {
            if (moment == null) named += 1;
            moment = carried;
        } else if (std.mem.eql(u8, name, body_scheme_part)) {
            if (body == null) named += 1;
            body = carried;
        } else if (std.mem.eql(u8, name, headers_scheme_part)) {
            if (headers == null) named += 1;
            headers = carried;
        } else if (std.mem.eql(u8, name, covered_headers_part)) {
            if (covered == null) named += 1;
            covered = carried;
        } else {
            named += 1;
        }
    }

    if (named < 2) return error.Unreadable;

    if (body) |written| read.body_code = try codeOf(written);
    if (headers) |written| read.headers_code = try codeOf(written);
    if (read.body_code == null and read.headers_code == null) return error.Unreadable;

    read.timestamp = try timestampOf(moment orelse return error.Unreadable);

    if (covered) |written| {
        if (written.len > 0) {
            var names = std.mem.splitScalar(u8, written, header_name_separator);
            while (names.next()) |name| {
                if (read.covered_count == max_covered_headers) return error.Unreadable;
                if (!isHeaderName(name)) return error.Unreadable;
                read.covered_headers[read.covered_count] = name;
                read.covered_count += 1;
            }
        }
    }

    return read;
}

/// The value that header of the delivery carried, under the name a signature refers to it by.
///
/// A later value wins over an earlier one under the same name, which is what a table built by the
/// caller would have done.
fn delivered(headers: []const Header, name: []const u8) ?[]const u8 {
    var found: ?[]const u8 = null;
    for (headers) |header| {
        if (std.ascii.eqlIgnoreCase(header.name, name)) found = header.value;
    }
    return found;
}

/// Whether the code this signature carries is the one the secret produces.
///
/// The stronger scheme wins when both are offered, and the comparison is made in constant time: one
/// that gave up at the first differing byte would say, by how long it took, how much of a guess was
/// right.
pub fn matches(
    read: *const Signature,
    payload: []const u8,
    headers: []const Header,
    subscription_secret: []const u8,
) SignatureError!bool {
    var code: Hmac = .init(subscription_secret);

    var moment: [32]u8 = undefined;
    const written = std.fmt.bufPrint(&moment, "{d}", .{read.timestamp}) catch return error.Unreadable;
    code.update(written);
    code.update(&.{message_separator});

    var out: [Hmac.mac_length]u8 = undefined;
    if (read.headers_code) |offered| {
        for (read.covered(), 0..) |name, index| {
            if (index > 0) code.update(&.{header_name_separator});
            code.update(name);
        }
        code.update(&.{message_separator});
        for (read.covered(), 0..) |name, index| {
            if (index > 0) code.update(&.{message_separator});
            code.update(delivered(headers, name) orelse return error.HeaderNotDelivered);
        }
        code.update(&.{message_separator});
        code.update(payload);
        code.final(&out);
        return std.crypto.timing_safe.eql([Hmac.mac_length]u8, out, offered);
    }

    // A signature carrying neither code is refused while it is being read, so what is left here is
    // the body-only scheme.
    code.update(payload);
    code.final(&out);
    return std.crypto.timing_safe.eql([Hmac.mac_length]u8, out, read.body_code.?);
}

/// Verifies a webhook against a moment the caller names.
///
/// The clock window is bilateral. A moment too far in the future is refused exactly like one too far
/// in the past, so the window a given delivery is accepted in stays the width the caller asked for,
/// whichever way a clock drifted.
///
/// `tolerance_seconds` is how far, in either direction, the moment the signature names may sit from
/// `current_time`. Five minutes is a reasonable trade-off between tolerating clock drift and bounding
/// how long a captured delivery can be replayed.
pub fn verifyWithCurrentTime(
    header: []const u8,
    payload: []const u8,
    headers: []const Header,
    subscription_secret: []const u8,
    tolerance_seconds: i64,
    current_time: i64,
) SignatureError!void {
    const read = try parse(header);

    // A header the signature covers that the request did not carry is settled before any code is
    // computed: signing over an absent value would let a sender drop a header and keep the
    // signature valid.
    for (read.covered()) |name| {
        if (delivered(headers, name) == null) return error.HeaderNotDelivered;
    }

    if (!try matches(&read, payload, headers, subscription_secret)) return error.CodeMismatch;

    const drift = current_time - read.timestamp;
    if (drift > tolerance_seconds or -drift > tolerance_seconds) return error.OutsideTolerance;
}

/// Verifies a webhook against the current moment.
///
/// The moment comes from the `Io` the caller hands in rather than from a clock this package reaches
/// on its own, which is what makes the window something a test can move.
pub fn verify(
    io: std.Io,
    header: []const u8,
    payload: []const u8,
    headers: []const Header,
    subscription_secret: []const u8,
    tolerance_seconds: i64,
) SignatureError!void {
    const now = std.Io.Timestamp.now(io, .real).toSeconds();
    return verifyWithCurrentTime(header, payload, headers, subscription_secret, tolerance_seconds, now);
}
